From 86fb73623f80b6549c46d5b2af35fa6390070c20 Mon Sep 17 00:00:00 2001
From: Dan Smith <danms@us.ibm.com>
Date: Wed, 22 Aug 2012 13:54:35 -0400
Subject: [PATCH 07/10] Fix exception handling in libvirt attach_volume()

Currently, the real reason for a failure is dropped when attempting
to unroll the volume connection (with the exception of when it is
VIR_ERR_OPERATION_FAILED). This change uses save_and_reraise_exception()
to correct that so that the actual reason for failure gets logged.

Fixes bug 1029463

Change-Id: Id47db565c4fb5a88d1a263600b41706dd3419726
Signed-off-by: Aaron Rosen <arosen@nicira.com>
---
 Authors                         |  1 +
 nova/virt/libvirt/connection.py | 12 +++++++-----
 2 files changed, 8 insertions(+), 5 deletions(-)

diff --git a/Authors b/Authors
index 4d784ca..c429dc3 100644
--- a/Authors
+++ b/Authors
@@ -39,6 +39,7 @@ Cole Robinson <crobinso@redhat.com>
 Cor Cornelisse <cor@hyves.nl>
 Cory Wright <corywright@gmail.com>
 Dan Prince <dprince@redhat.com>
+Dan Smith <danms@us.ibm.com>
 Dan Wendlandt <dan@nicira.com>
 Daniel P. Berrange <berrange@redhat.com>
 Dave Lapsley <dlapsley@nicira.com>
diff --git a/nova/virt/libvirt/connection.py b/nova/virt/libvirt/connection.py
index 24112da..238ba3a 100644
--- a/nova/virt/libvirt/connection.py
+++ b/nova/virt/libvirt/connection.py
@@ -537,15 +537,17 @@ class LibvirtConnection(driver.ComputeDriver):
             try:
                 virt_dom.attachDevice(xml)
             except Exception, ex:
-                self.volume_driver_method('disconnect_volume',
-                                           connection_info,
-                                           mount_device)
-
                 if isinstance(ex, libvirt.libvirtError):
                     errcode = ex.get_error_code()
                     if errcode == libvirt.VIR_ERR_OPERATION_FAILED:
+                        self.volume_driver_method('disconnect_volume',
+                                                   connection_info,
+                                                   mount_device)
                         raise exception.DeviceIsBusy(device=mount_device)
-                raise
+                with utils.save_and_reraise_exception():
+                    self.volume_driver_method('disconnect_volume',
+                                               connection_info,
+                                               mount_device)
 
     @staticmethod
     def _get_disk_xml(xml, device):
-- 
1.7.11.4

